### PR TITLE
brainbuilder → 0.9.1; placement-algorithm → 2.0.1

### PR DESCRIPTION
**Pull request by arsenius7**

### CODE REVIEW

File: deploy/packages/bbp-packages.yaml [hunk 1]
```diff
@@ -46,7 +46,7 @@ packages:
       #- python
     specs:
       - parquet-converters@0.3
-      - placement-algorithm@2.0.0
+      - placement-algorithm@2.0.1%gcc@8.2.0
```
Comment (arsenius7) [inline deploy/packages/bbp-packages.yaml]: I have noticed that with `%gcc@8.2.0` the resulting module works ~30% faster, and that's a speedup that would be nice to have considering that `placement-algorithm` run takes hours for 10M circuit.
Would be interesting to understand the reason for the difference in the performance, but I have not investigated it yet.

Comment (arsenius7) [reply deploy/packages/bbp-packages.yaml]: CI is failing because of `%gcc@8.2.0` as well.
Apparently once again I'm doing something different locally. ^^

Comment (matz-e) [reply deploy/packages/bbp-packages.yaml]: @arsenius7 the compiler here is defined by `gnu-stable` a few lines above. Let me have a look.

**Review by matz-e — COMMENTED** [review]
Compiler/infrastructure setup is happening in `packages/toolchains.yaml`. You can see there that `gnu-stable` is configured to use `gcc@6.4.0`, hence your build fail.

**Comment by matz-e** [comment]
Please retest

**Comment by matz-e** [comment]
Please retest

Seems like something has crashed?

**Comment by matz-e** [comment]
@arsenius7 this is how we'd add a new compiler. Did you compile your whole stack with the new gcc when running locally?

**Comment by arsenius7** [comment]
I ran
```
$ spack spec placement-algorithm%gcc@8.2.0
```
which gave me:
```
Input spec
--------------------------------
placement-algorithm%gcc@8.2.0

Concretized
--------------------------------
placement-algorithm@2.0.1%gcc@8.2.0 arch=linux-rhel7-x86_64 
    ^py-lxml@4.3.0%gcc@8.2.0 arch=linux-rhel7-x86_64 
        ^libxml2@2.9.8%gcc@8.2.0~python arch=linux-rhel7-x86_64 
        ^libxslt@1.1.29%gcc@8.2.0+crypto arch=linux-rhel7-x86_64 
        ^py-cython@0.29%gcc@8.2.0 arch=linux-rhel7-x86_64 
            ^python@3.6.5%gcc@8.2.0+dbm~optimizations patches=123082ab3483ded78e86d7c809e98a804b3465b4683c96bd79a2fd799f572244 +pic+pythoncmd+shared~tk~ucs4 arch=linux-rhel7-x86_64 
        ^py-setuptools@40.4.3%gcc@8.2.0 arch=linux-rhel7-x86_64 
    ^py-morph-tool@0.1.0%gcc@8.2.0~neuron arch=linux-rhel7-x86_64 
        ^py-click@7.0%gcc@8.2.0 arch=linux-rhel7-x86_64 
        ^py-morphio@2.0.3%gcc@8.2.0 arch=linux-rhel7-x86_64 
            ^cmake@3.13.0%gcc@8.2.0~doc+ncurses+openssl+ownlibs~qt arch=linux-rhel7-x86_64 
            ^py-numpy@1.15.2%gcc@8.2.0+blas+lapack arch=linux-rhel7-x86_64 
                ^intel-mkl@2018.1.163%gcc@8.2.0~ilp64+shared threads=none arch=linux-rhel7-x86_64 
    ^py-mpi4py@3.0.0%gcc@8.2.0 arch=linux-rhel7-x86_64 
        ^hpe-mpi@2.16%gcc@8.2.0 arch=linux-rhel7-x86_64 
    ^py-pandas@0.23.4%gcc@8.2.0 arch=linux-rhel7-x86_64 
        ^py-bottleneck@1.2.1%gcc@8.2.0 arch=linux-rhel7-x86_64 
        ^py-dateutil@2.5.2%gcc@8.2.0 arch=linux-rhel7-x86_64 
            ^py-six@1.11.0%gcc@8.2.0 arch=linux-rhel7-x86_64 
        ^py-numexpr@2.6.5%gcc@8.2.0 arch=linux-rhel7-x86_64 
        ^py-pytz@2017.2%gcc@8.2.0 arch=linux-rhel7-x86_64 
    ^py-tqdm@4.8.4%gcc@8.2.0 arch=linux-rhel7-x86_64 
    ^py-ujson@1.35%gcc@8.2.0 arch=linux-rhel7-x86_64 
    ^py-voxcell@2.5.5%gcc@8.2.0 arch=linux-rhel7-x86_64 
        ^py-future@0.16.0%gcc@8.2.0 arch=linux-rhel7-x86_64 
        ^py-h5py@2.8.0%gcc@8.2.0~mpi arch=linux-rhel7-x86_64 
            ^hdf5@1.10.4%gcc@8.2.0~cxx~debug~fortran+hl~mpi+pic+shared~szip~threadsafe arch=linux-rhel7-x86_64 
                ^zlib@1.2.11%gcc@8.2.0+optimize+pic+shared arch=linux-rhel7-x86_64 
            ^py-pkgconfig@1.2.2%gcc@8.2.0 arch=linux-rhel7-x86_64 
                ^pkg-config@0.29.2%gcc@8.2.0+internal_glib arch=linux-rhel7-x86_64 
                ^py-nose@1.3.7%gcc@8.2.0 arch=linux-rhel7-x86_64 
        ^py-pynrrd@0.2.5%gcc@8.2.0 arch=linux-rhel7-x86_64 
        ^py-requests@2.21.0%gcc@8.2.0 arch=linux-rhel7-x86_64 
            ^py-certifi@2018.11.29%gcc@8.2.0 arch=linux-rhel7-x86_64 
            ^py-chardet@3.0.4%gcc@8.2.0 arch=linux-rhel7-x86_64 
            ^py-idna@2.5%gcc@8.2.0 arch=linux-rhel7-x86_64 
            ^py-urllib3@1.24%gcc@8.2.0 arch=linux-rhel7-x86_64 
        ^py-scipy@1.1.0%gcc@8.2.0 arch=linux-rhel7-x86_64 
```
then
```
spack install placement-algorithm%gcc@8.2.0
```
and tested the result module.
I have to admit that I'm working with Spack a bit at hazard (which I find a good sign of intuitive system :) ).

**Comment by pramodk** [comment]
@arsenius7  : 

> spack spec placement-algorithm%gcc@8.2.0

In order to use newer gcc, we first have to make sure the compiler is available to Spack (compilers are not dependencies yet and hence those are not installed as part of `spack install`).

On your local system, you might have gcc@8 available (or installed via spack). 

By the way, which linux distribution you use? (Ubuntu@18.10?).

In the future we can easily setup CI with docker image ([like here](https://github.com/pramodk/neuron-docker-spack/blob/master/Dockerfile)) to make sure we can have consistent software stack buildable on our desktop (without errors).

File: deploy/packages/bbp-packages.yaml [hunk 2]
```diff
@@ -39,14 +39,15 @@ packages:
   gnu-stable-parallel-python:
     target_matrix:
       - gnu-stable
```
Comment (matz-e) [inline deploy/packages/bbp-packages.yaml]:
```suggestion
```
Since both packages here compile with 8.2.0, `gnu-stable` could be removed.

**Comment by arsenius7** [comment]
I tried it on BB5 by blindly appending `%gcc@8.2.0` as described above.

As I've pointed out, I'm not sure what was the reason for the observed speedup, but knowing that it _can_ be achieved I'd interested in getting it.

I can revert `%gcc@8.2.0` part of the change for the moment, until it's investigated better.

**Review by matz-e — CHANGES_REQUESTED** [review]

**Comment by matz-e** [comment]
> I tried it on BB5 by blindly appending `%gcc@8.2.0` as described above.
> 
> As I've pointed out, I'm not sure what was the reason for the observed speedup, but knowing that it _can_ be achieved I'd interested in getting it.
> 
> I can revert `%gcc@8.2.0` part of the change for the moment, until it's investigated better.

I've pushed changes that do the setup/compilation with `%gcc@8.2.0` properly. I'd be interested in having 8.x.y at least in the testing pipeline, if @pramodk is OK with it.

**Comment by pramodk** [comment]
> I'd be interested in having 8.x.y at least in the testing pipeline, if @pramodk is OK with it.

Its perfectly fine to test `8.x`. The only reason we didn't move to recent gcc (at least 7.x) was all teams are still with 6.4.0 (from nix). We can obviously deploy all libraries with gcc-8 as well but then archiving and size on SSD is another aspect we have to consider.

Anyway, I expect this deploy discussion (& decision) to conclude in coming weeks and then we can agree on next major gcc version to use. (from hpc side we are ready for v8 but need to validate some stuff)

**Comment by pramodk** [comment]
And another aspect for consistency is intel and pgi compiler....they are currently using gcc/6.4.0 libstdc++ underneath. If want to combine libraries between compilers (not in this case), it's not bad idea to keep consistent compilers (if we can!)

**Comment by matz-e** [comment]
Let me check which GCC version the modules pick up

**Review by matz-e — APPROVED** [review]
LGTM now. Sorry for mucking around with the PR so much!

**Comment by arsenius7** [comment]
Np; in a sense I have shifted the job of `%gcc@8.2.0` on your shoulders here. ^^

**Comment by matz-e** [comment]
Please retest

There's some module confusion going on :)

**Comment by matz-e** [comment]
Well, at least 8.2.0 has `std::filesystem`… that's a good step forward, and additionally more helpful error messages!

**Comment by pramodk** [comment]
Please retest

**Review by pramodk — APPROVED** [review]

**Pull request merged**